File: georocket-server/src/main/java/io/georocket/index/xml/GmlIdIndexerFactory.java
package io.georocket.index.xml;

import java.util.Map;

import com.google.common.collect.ImmutableMap;

import io.georocket.query.ElasticsearchQueryHelper;
import io.georocket.query.KeyValueQueryPart;
import io.georocket.query.KeyValueQueryPart.ComparisonOperator;
import io.georocket.query.QueryPart;
import io.georocket.query.StringQueryPart;
import io.vertx.core.json.JsonObject;

/**
 * Create instances of {@link GmlIdIndexer}
 * @author Michel Kraemer
 */
public class GmlIdIndexerFactory implements XMLIndexerFactory {
  @Override
  public XMLIndexer createIndexer() {
    return new GmlIdIndexer();
  }
  
  @Override
  public Map<String, Object> getMapping() {
    return ImmutableMap.of("properties", ImmutableMap.of("gmlIds", ImmutableMap.of(
        "type", "keyword" // array of keywords actually, auto-supported by Elasticsearch
    )));
  }

  /**
   * Test if the given key-value query part refers to a gmlId and if it uses
   * the EQ operator (e.g. EQ(gmlId myId) or EQ(gml:id myId))
   * @param kvqp the key-value query part to check
   * @return true if it refers to a gmlId, false otherwise
   */
  private static boolean isGmlIdEQ(KeyValueQueryPart kvqp) {
    String key = kvqp.getKey();
    ComparisonOperator comp = kvqp.getComparisonOperator();
    return (comp == ComparisonOperator.EQ && ("gmlId".equals(key) || "gml:id".equals(key)));
  }

  @Override
  public MatchPriority getQueryPriority(QueryPart queryPart) {
    if (queryPart instanceof StringQueryPart) {
      return MatchPriority.SHOULD;
    }
    if (queryPart instanceof KeyValueQueryPart && isGmlIdEQ((KeyValueQueryPart)queryPart)) {
      return MatchPriority.SHOULD;
    }
    return MatchPriority.NONE;
  }

  @Override
  public JsonObject compileQuery(QueryPart queryPart) {
    if (queryPart instanceof StringQueryPart) {
      String search = ((StringQueryPart)queryPart).getSearchString();
      return ElasticsearchQueryHelper.termQuery("gmlIds", search);
    } else if (queryPart instanceof KeyValueQueryPart) {
      KeyValueQueryPart kvqp = (KeyValueQueryPart)queryPart;
      if (isGmlIdEQ(kvqp)) {
        return ElasticsearchQueryHelper.termQuery("gmlIds", kvqp.getValue());
      }
    }
    return null;
  }
}

File: georocket-server/src/main/kotlin/io/georocket/index/xml/GmlIdIndexerFactory.kt
package io.georocket.index.xml

import io.georocket.index.Indexer
import io.georocket.index.IndexerFactory
import io.georocket.query.QueryCompiler.MatchPriority
import io.georocket.query.QueryPart
import io.georocket.query.QueryPart.ComparisonOperator
import io.georocket.query.StringQueryPart
import io.georocket.util.StreamEvent
import io.georocket.util.XMLStreamEvent
import io.vertx.core.json.JsonObject
import io.vertx.kotlin.core.json.jsonObjectOf

/**
 * Create instances of [GmlIdIndexer]
 * @author Michel Kraemer
 */
class GmlIdIndexerFactory : IndexerFactory {
  override fun <T : StreamEvent> createIndexer(eventType: Class<T>): Indexer<T>? {
    if (eventType.isAssignableFrom(XMLStreamEvent::class.java)) {
      @Suppress("UNCHECKED_CAST")
      return GmlIdIndexer() as Indexer<T>
    }
    return null
  }

  /**
   * Test if the given key-value query part refers to a gmlId and if it uses
   * the EQ operator (e.g. EQ(gmlId myId) or EQ(gml:id myId))
   */
  private fun isGmlIdEQ(qp: StringQueryPart): Boolean {
    val key = qp.key
    val comp = qp.comparisonOperator
    return comp === ComparisonOperator.EQ && ("gmlId" == key || "gml:id" == key)
  }

  override fun getQueryPriority(queryPart: QueryPart): MatchPriority {
    return if (queryPart is StringQueryPart) {
      if (queryPart.key == null || isGmlIdEQ(queryPart)) {
        return MatchPriority.SHOULD
      } else {
        MatchPriority.NONE
      }
    } else {
      MatchPriority.NONE
    }
  }

  override fun compileQuery(queryPart: QueryPart): JsonObject? {
    return if (queryPart is StringQueryPart) {
      if (queryPart.key == null || isGmlIdEQ(queryPart)) {
        jsonObjectOf("gmlIds" to queryPart.value)
      } else {
        null
      }
    } else {
      null
    }
  }
}
